rfcs/framework-0035-error-handling-and-response.md

```diff
@@ -0,0 +1,323 @@
+
+1. Flask layer, for instance routing errors.
+1. View layer, for instance ``abort(400)`` or ``raise BadRequest()``.
+1. Service/data source layer, for instance Elasticsearch ``RequestError`` or ``PIDDeleteError``.
+
+A bit oversimplified, you can say that Flask/view layer errors usaully  result in HTTP errors such as 40x errors (Bad Request, Not Found).
+
+Service/data source layer errors will usually result in 500 Internal Server Error *unless they are handled*.
+
+For all above type of errors, we'd like to create a machine readable error response.
+
+**View error handling**
+
+A Flask view can simply catch exceptions, and render a response. This can either be done per view or via special decorators.
+
+```python
+@app.route('/')
+def view1():
+    try:
+        # ...
+    except Exception:
+        return 400, ...
+```
+
+The view layer error handling is best suited for very custom error handling from the service layer. E.g implementation of a custom API with custom reposnses.
+
+The issue with view layer error handling is that only errors from the view and service/data source layer is being caught. Any error in the Flask layer (routing, request processing functions etc) will escape the try/except-clause (because they happen outside the function).
+
+Example of a 404 error:
+
+1. A routing error will result in a 404 error from the Flask layer.
+2. An object that doesn't exists in the database, will likely result in a 404 error from the service layer.
+
+A view layer can only catch the latter. Thus, if we render an error response in the view, we still have to deal with error response for Flask layer errors.
+
+**Flask's error handling/rendering**
+
+Flask and Werkzeug's request dispatching system will catch all errors happening in the three layers described. The error handling is pretty sophisticated, and is explained below.
+
+First, Flask will look for a user-provided error handler. If found, the handler is responsible for making a response. If not found, but the exception is an HTTPException (or subclass thereof), the HTTPException will simply be returned, because an HTTPException can self-render. If it's not an HTTPException, the error is simply reraised and results in an Internal Server Error. The following pseudo-code is an example of the handling:
+
+```python
+def handle_exception(e):
+    # If it's an HTTPException, use different method:
+    if isinstance(e, HTTPException):
+        return handle_http_exception(e)
+
+    handler = find_error_handler(e)
+    if handler is None:
+        # No handler was found, and it's not and HTTPException.
+        raise # <----- reraise
+    # A handler was found, use it to return response
+    return handler(e)
+
+def handle_http_exception(e):
+    handler = find_error_handler(e)
+    if handler is None:
+        # No handler was found, but it is an HTTPException.
+        return e # <----- return
+    # A handler was found, use it to return response
+    return handler(e)
```

Comment:
Not sure I follow why HttpExceptions are treated differently. The pseudo-code above simplifies to:

 ```suggestion
def handle_exception(e):
    handler = find_error_handler(e)
    if handler is None:
        # No handler was found
        raise e  # <----- reraise. This results in a 500 if e is not an HTTPException or 
                 # the appropriate status code because it is an HTTPException
    # A handler was found, use it to return response
    raise handler(e)

```

Reply:
It's because the "pseudo-code" is taken directly from Flask were there are two methods, so I prefer to keep it as close as possible to Flask code.